src/eth_provider/error.rs

```diff
@@ -71,3 +73,149 @@ impl From<EthProviderError> for ErrorObject<'static> {
 pub fn rpc_err(code: EthRpcErrorCode, msg: impl Into<String>) -> jsonrpsee::types::error::ErrorObject<'static> {
     jsonrpsee::types::error::ErrorObject::owned(code as i32, msg.into(), None::<()>)
 }
+
+/// Error related to the Kakarot eth provider
+/// which utilizes the starknet provider and
+/// a database internally.
+#[derive(Debug, Error)]
+pub enum KakarotError {
+    /// Error related to the starknet provider.
+    #[error(transparent)]
+    ProviderError(#[from] starknet::providers::ProviderError),
+    /// Error related to the database.
+    #[error(transparent)]
+    DatabaseError(#[from] mongodb::error::Error),
+    /// Error related to the evm execution.
+    #[error(transparent)]
+    ExecutionError(EvmError),
+    /// Error related to a starknet call.
+    #[error(transparent)]
+    CallError(#[from] cainome::cairo_serde::Error),
+    /// Error related to starknet to eth conversion or vice versa.
+    #[error(transparent)]
+    ConversionError(#[from] ConversionError),
+}
+
+impl From<KakarotError> for EthApiError {
+    fn from(value: KakarotError) -> Self {
+        EthApiError::Internal(value)
+    }
+}
+
+/// Error related to EVM execution.
+#[derive(Debug, Error)]
+pub enum EvmError {
+    #[error("validation failed")]
+    ValidationError,
+    #[error("state modification error")]
+    StateModificationError,
+    #[error("unknown opcode")]
+    UnknownOpcode,
+    #[error("invalid jump dest")]
+    InvalidJumpDest,
+    #[error("invalid caller")]
+    NotKakarotEoaCaller,
+    #[error("view function error")]
+    ViewFunctionError,
+    #[error("stack overflow")]
+    StackOverflow,
+    #[error("stack underflow")]
+    StackUnderflow,
+    #[error("out of bounds read")]
+    OutOfBoundsRead,
+    #[error("unknown precompile {0}")]
+    UnknownPrecompile(String),
+    #[error("not implemented precompile {0}")]
+    NotImplementedPrecompile(String),
+    #[error("precompile input error")]
+    PrecompileInputError,
+    #[error("precompile flag error")]
+    PrecompileFlagError,
+    #[error("balance error")]
+    BalanceError,
+    #[error("address collision")]
+    AddressCollision,
+    #[error("out of gas")]
+    OutOfGas,
+    #[error("{0}")]
+    Other(String),
+}
+
+impl From<EvmError> for KakarotError {
+    fn from(value: EvmError) -> Self {
+        KakarotError::ExecutionError(value)
+    }
+}
+
+impl From<String> for EvmError {
```

Comment:
How can we test this extensively?

Reply:
other then unit testing there isn't really a way to do it

Reply:
how about integration test? sending txs that'll fail in those ways?